src/box/memtx_rtree.cc

```diff
@@ -146,20 +146,139 @@ extract_rectangle(struct rtree_rect *rect, struct tuple *tuple,
 /* {{{ MemtxRTree Iterators ****************************************/
 
 struct index_rtree_iterator {
-        struct iterator base;
-        struct rtree_iterator impl;
+	struct iterator base;
+	struct rtree_iterator impl;
 	/** Memory pool the iterator was allocated from. */
 	struct mempool *pool;
+	/* last fetched tuple (pos). Is used for pagination */
+	struct tuple *last;
 };
 
+/* Min distance of last saved tuple.
+ * Is used for pagintion optimisation.
+ * Euclid distance.
+ */
+static sq_coord_t
+rtree_rect_neigh_distance_pos(
+	const struct rtree_rect *rect,
+	const struct rtree_rect *neigh_rect,
+	unsigned dimension)
+{
+	sq_coord_t result = 0, result2 = 0;
+	size_t counter = 0;
+	for (int i = dimension; --i >= 0; ) {
+		const coord_t *coords = &rect->coords[2 * i];
+		coord_t neigh_coord = neigh_rect->coords[2 * i];
+		if (neigh_coord < coords[0]) {
+			sq_coord_t diff = (sq_coord_t)(coords[0] - neigh_coord);
+			result += diff * diff;
+		} else if (neigh_coord > coords[1]) {
+			sq_coord_t diff = (sq_coord_t)(neigh_coord - coords[1]);
+			result += diff * diff;
+		} else if ((neigh_coord - coords[0]) <
+				(coords[1] - neigh_coord)) {
+			counter++;
+			sq_coord_t diff = (sq_coord_t)(neigh_coord - coords[0]);
+			if (!result2 || result2 > diff * diff)
+				result2 = diff * diff;
+		} else {
+			counter++;
+			sq_coord_t diff = (sq_coord_t)(neigh_coord - coords[1]);
+			if (!result2 || result2 > diff * diff)
+				result2 = diff * diff;
+		}
+	}
+	/* if point is in rect */
+	if (counter == dimension)
+		result = result2;
+	return result;
+}
+
+/* Min distance of last saved tuple.
+ * Is used for pagintion optimisation.
+ * Manhattan distance
+ */
+static sq_coord_t
+rtree_rect_neigh_distance_pos2(
+	const struct rtree_rect *rect,
+	const struct rtree_rect *neigh_rect,
+	unsigned dimension)
+{
+	sq_coord_t result = 0, result2 = 0;
+	size_t counter = 0;
+	for (int i = dimension; --i >= 0; ) {
+		const coord_t *coords = &rect->coords[2 * i];
+		coord_t neigh_coord = neigh_rect->coords[2 * i];
+		if (neigh_coord < coords[0]) {
```

Comment:
Let's factor this logic of choosing diff out of the function. The function would return `diff` and flag `is_in_rectangle`.

Also, please, cover the new function with an explanation of what it does.